File: packages/KNIRVSDK/go/signing/wasm_test.go
package signing

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

type wasmGoldenVectors struct {
	WasmPublicationV1 struct {
		Input struct {
			NetworkID           string `json:"network_id"`
			NetworkFingerprint  string `json:"network_fingerprint"`
			ArtifactDigest      string `json:"artifact_digest"`
			ByteSize            uint64 `json:"byte_size"`
			ModuleKind          string `json:"module_kind"`
			AbiVersion          uint64 `json:"abi_version"`
			ModuleSchemaVersion uint64 `json:"module_schema_version"`
			BuildID             string `json:"build_id"`
			ToolchainDigest     string `json:"toolchain_digest"`
			SelfTestDigest      string `json:"self_test_digest"`
			PublisherAddress    string `json:"publisher_address"`
			DveTemplateID       string `json:"dve_template_id"`
		} `json:"input"`
		CanonicalHex string `json:"canonical_hex"`
	} `json:"wasm_publication_v1"`
	RelayEnvelopeV1 struct {
		Input struct {
			RequestID     string `json:"request_id"`
			UserSubject   string `json:"user_subject"`
			DeviceID      string `json:"device_id"`
			DveID         string `json:"dve_id"`
			TargetType    string `json:"target_type"`
			TargetID      string `json:"target_id"`
			Capability    string `json:"capability"`
			Sequence      uint64 `json:"sequence"`
			LeaseEpoch    uint64 `json:"lease_epoch"`
			IssuedAtUnix  int64  `json:"issued_at_unix"`
			ExpiresAtUnix int64  `json:"expires_at_unix"`
			PayloadDigest string `json:"payload_digest"`
		} `json:"input"`
		CanonicalHex string `json:"canonical_hex"`
	} `json:"relay_envelope_v1"`
	WasmManifestV1 struct {
		Input struct {
			ManifestID         string `json:"manifest_id"`
			NetworkID          string `json:"network_id"`
			ChainID            string `json:"chain_id"`
			NetworkFingerprint string `json:"network_fingerprint"`
			UserSubject        string `json:"user_subject"`
			DeviceID           string `json:"device_id"`
			DveID              string `json:"dve_id"`
			LeaseEpoch         uint64 `json:"lease_epoch"`
			Modules            []struct {
				ModuleKind                 string `json:"module_kind"`
				ArtifactDigest             string `json:"artifact_digest"`
				ByteSize                   uint64 `json:"byte_size"`
				AbiVersion                 uint64 `json:"abi_version"`
				ModuleSchemaVersion        uint64 `json:"module_schema_version"`
				CapabilitiesJSON           string `json:"capabilities_json"`
				ConfigurationDigest        string `json:"configuration_digest"`
				DownloadPath               string `json:"download_path"`
				PublisherAddress           string `json:"publisher_address"`
				PublicationStatementDigest string `json:"publication_statement_digest"`
			} `json:"modules"`
			RelayTargetType        string `json:"relay_target_type"`
			RelayTargetID          string `json:"relay_target_id"`
			AssignmentID           string `json:"assignment_id"`
			AssignmentVersion      uint64 `json:"assignment_version"`
			SupersedesAssignmentID string `json:"supersedes_assignment_id"`
		} `json:"input"`
		CanonicalHex string `json:"canonical_hex"`
	} `json:"wasm_manifest_v1"`
}

func loadWasmGoldenVectors(t *testing.T) wasmGoldenVectors {
	t.Helper()
	data, err := os.ReadFile("../../testvectors/wasm_payloads.json")
	if err != nil {
		t.Fatalf("failed to read golden vectors: %v", err)
	}
	var vectors wasmGoldenVectors
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("failed to parse golden vectors: %v", err)
	}
	return vectors
}

func TestWasmPublicationPayloadGoldenVector(t *testing.T) {
	vectors := loadWasmGoldenVectors(t)
	in := vectors.WasmPublicationV1.Input

	payload := WasmPublicationPayload{
		NetworkID:           in.NetworkID,
		NetworkFingerprint:  in.NetworkFingerprint,
		ArtifactDigest:      in.ArtifactDigest,
		ByteSize:            in.ByteSize,
		ModuleKind:          in.ModuleKind,
		AbiVersion:          in.AbiVersion,
		ModuleSchemaVersion: in.ModuleSchemaVersion,
		BuildID:             in.BuildID,
		ToolchainDigest:     in.ToolchainDigest,
		SelfTestDigest:      in.SelfTestDigest,
		PublisherAddress:    in.PublisherAddress,
		DveTemplateID:       in.DveTemplateID,
	}

	got, err := MarshalWasmPublicationPayload(payload)
	if err != nil {
		t.Fatalf("MarshalWasmPublicationPayload: %v", err)
	}
	want, err := hex.DecodeString(vectors.WasmPublicationV1.CanonicalHex)
	if err != nil {
		t.Fatalf("decode golden hex: %v", err)
	}
	if hex.EncodeToString(got) != hex.EncodeToString(want) {
		t.Fatalf("canonical bytes mismatch:\n got=%s\nwant=%s", hex.EncodeToString(got), hex.EncodeToString(want))
	}

	roundTrip, err := ParseWasmPublicationPayload(got)
	if err != nil {
		t.Fatalf("ParseWasmPublicationPayload: %v", err)
	}
	if roundTrip.ArtifactDigest != in.ArtifactDigest || roundTrip.PublisherAddress != in.PublisherAddress || roundTrip.ByteSize != in.ByteSize {
		t.Fatalf("round trip mismatch: %+v", roundTrip)
	}
}

func TestRelayEnvelopeGoldenVector(t *testing.T) {
	vectors := loadWasmGoldenVectors(t)
	in := vectors.RelayEnvelopeV1.Input

	envelope := RelayEnvelope{
		RequestID:     in.RequestID,
		UserSubject:   in.UserSubject,
		DeviceID:      in.DeviceID,
		DveID:         in.DveID,
		TargetType:    in.TargetType,
		TargetID:      in.TargetID,
		Capability:    in.Capability,
		Sequence:      in.Sequence,
		LeaseEpoch:    in.LeaseEpoch,
		IssuedAtUnix:  in.IssuedAtUnix,
		ExpiresAtUnix: in.ExpiresAtUnix,
		PayloadDigest: in.PayloadDigest,
	}

	got, err := MarshalRelayEnvelope(envelope)
	if err != nil {
		t.Fatalf("MarshalRelayEnvelope: %v", err)
	}
	want, err := hex.DecodeString(vectors.RelayEnvelopeV1.CanonicalHex)
	if err != nil {
		t.Fatalf("decode golden hex: %v", err)
	}
	if hex.EncodeToString(got) != hex.EncodeToString(want) {
		t.Fatalf("canonical bytes mismatch:\n got=%s\nwant=%s", hex.EncodeToString(got), hex.EncodeToString(want))
	}

	roundTrip, err := ParseRelayEnvelope(got)
	if err != nil {
		t.Fatalf("ParseRelayEnvelope: %v", err)
	}
	if roundTrip.RequestID != in.RequestID || roundTrip.Sequence != in.Sequence || roundTrip.PayloadDigest != in.PayloadDigest {
		t.Fatalf("round trip mismatch: %+v", roundTrip)
	}
}

func TestWasmManifestPayloadGoldenVector(t *testing.T) {
	vectors := loadWasmGoldenVectors(t)
	in := vectors.WasmManifestV1.Input

	modules := make([]WasmManifestModule, len(in.Modules))
	for i, m := range in.Modules {
		modules[i] = WasmManifestModule{
			ModuleKind:                 m.ModuleKind,
			ArtifactDigest:             m.ArtifactDigest,
			ByteSize:                   m.ByteSize,
			AbiVersion:                 m.AbiVersion,
			ModuleSchemaVersion:        m.ModuleSchemaVersion,
			CapabilitiesJSON:           m.CapabilitiesJSON,
			ConfigurationDigest:        m.ConfigurationDigest,
			DownloadPath:               m.DownloadPath,
			PublisherAddress:           m.PublisherAddress,
			PublicationStatementDigest: m.PublicationStatementDigest,
		}
	}

	payload := WasmManifestPayload{
		ManifestID:             in.ManifestID,
		NetworkID:              in.NetworkID,
		ChainID:                in.ChainID,
		NetworkFingerprint:     in.NetworkFingerprint,
		UserSubject:            in.UserSubject,
		DeviceID:               in.DeviceID,
		DveID:                  in.DveID,
		LeaseEpoch:             in.LeaseEpoch,
		Modules:                modules,
		RelayTargetType:        in.RelayTargetType,
		RelayTargetID:          in.RelayTargetID,
		AssignmentID:           in.AssignmentID,
		AssignmentVersion:      in.AssignmentVersion,
		SupersedesAssignmentID: in.SupersedesAssignmentID,
	}

	got, err := MarshalWasmManifestPayload(payload)
	if err != nil {
		t.Fatalf("MarshalWasmManifestPayload: %v", err)
	}
	want, err := hex.DecodeString(vectors.WasmManifestV1.CanonicalHex)
	if err != nil {
		t.Fatalf("decode golden hex: %v", err)
	}
	if hex.EncodeToString(got) != hex.EncodeToString(want) {
		t.Fatalf("canonical bytes mismatch:\n got=%s\nwant=%s", hex.EncodeToString(got), hex.EncodeToString(want))
	}

	roundTrip, err := ParseWasmManifestPayload(got)
	if err != nil {
		t.Fatalf("ParseWasmManifestPayload: %v", err)
	}
	if len(roundTrip.Modules) != len(in.Modules) {
		t.Fatalf("expected %d modules, got %d", len(in.Modules), len(roundTrip.Modules))
	}
	if roundTrip.AssignmentID != in.AssignmentID || roundTrip.Modules[0].ArtifactDigest != in.Modules[0].ArtifactDigest {
		t.Fatalf("round trip mismatch: %+v", roundTrip)
	}
}

File: packages/KNIRVSDK/ts/unified/src/wasm.test.ts
import {
  marshalWasmPublicationPayload,
  parseWasmPublicationPayload,
  marshalWasmManifestPayload,
  parseWasmManifestPayload,
  WasmPublicationPayload,
  WasmManifestPayload,
} from './wasm';
import { marshalRelayEnvelope, parseRelayEnvelope, RelayEnvelope } from './relay';

// Cross-language golden vectors: KNIRV_NETWORK/packages/KNIRVSDK/testvectors/wasm_payloads.json.
// Both this file and go/signing/wasm_test.go must derive canonical_hex byte-for-byte from the
// same input — see KNIRV_CORP/packages/controller/stateless_pwa_controller.md section 3.4.
// eslint-disable-next-line @typescript-eslint/no-var-requires
const vectors = require('../../../testvectors/wasm_payloads.json');

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

describe('wasm canonical payloads — golden vectors', () => {
  it('matches the Go-derived wasm_publication_v1 canonical bytes', () => {
    const input = vectors.wasm_publication_v1.input;
    const payload: WasmPublicationPayload = {
      networkId: input.network_id,
      networkFingerprint: input.network_fingerprint,
      artifactDigest: input.artifact_digest,
      byteSize: input.byte_size,
      moduleKind: input.module_kind,
      abiVersion: input.abi_version,
      moduleSchemaVersion: input.module_schema_version,
      buildId: input.build_id,
      toolchainDigest: input.toolchain_digest,
      selfTestDigest: input.self_test_digest,
      publisherAddress: input.publisher_address,
      dveTemplateId: input.dve_template_id,
    };

    const bytes = marshalWasmPublicationPayload(payload);
    expect(toHex(bytes)).toBe(vectors.wasm_publication_v1.canonical_hex);

    const roundTrip = parseWasmPublicationPayload(bytes);
    expect(roundTrip.artifactDigest).toBe(input.artifact_digest);
    expect(roundTrip.publisherAddress).toBe(input.publisher_address);
    expect(String(roundTrip.byteSize)).toBe(String(input.byte_size));
  });

  it('matches the Go-derived wasm_manifest_v1 canonical bytes', () => {
    const input = vectors.wasm_manifest_v1.input;
    const payload: WasmManifestPayload = {
      manifestId: input.manifest_id,
      networkId: input.network_id,
      chainId: input.chain_id,
      networkFingerprint: input.network_fingerprint,
      userSubject: input.user_subject,
      deviceId: input.device_id,
      dveId: input.dve_id,
      leaseEpoch: input.lease_epoch,
      modules: input.modules.map((m: any) => ({
        moduleKind: m.module_kind,
        artifactDigest: m.artifact_digest,
        byteSize: m.byte_size,
        abiVersion: m.abi_version,
        moduleSchemaVersion: m.module_schema_version,
        capabilitiesJson: m.capabilities_json,
        configurationDigest: m.configuration_digest,
        downloadPath: m.download_path,
        publisherAddress: m.publisher_address,
        publicationStatementDigest: m.publication_statement_digest,
      })),
      relayTargetType: input.relay_target_type,
      relayTargetId: input.relay_target_id,
      assignmentId: input.assignment_id,
      assignmentVersion: input.assignment_version,
      supersedesAssignmentId: input.supersedes_assignment_id,
    };

    const bytes = marshalWasmManifestPayload(payload);
    expect(toHex(bytes)).toBe(vectors.wasm_manifest_v1.canonical_hex);

    const roundTrip = parseWasmManifestPayload(bytes);
    expect(roundTrip.modules).toHaveLength(input.modules.length);
    expect(roundTrip.assignmentId).toBe(input.assignment_id);
    expect(roundTrip.modules[0].artifactDigest).toBe(input.modules[0].artifact_digest);
  });

  it('matches the Go-derived relay_envelope_v1 canonical bytes', () => {
    const input = vectors.relay_envelope_v1.input;
    const envelope: RelayEnvelope = {
      requestId: input.request_id,
      userSubject: input.user_subject,
      deviceId: input.device_id,
      dveId: input.dve_id,
      targetType: input.target_type,
      targetId: input.target_id,
      capability: input.capability,
      sequence: input.sequence,
      leaseEpoch: input.lease_epoch,
      issuedAtUnix: input.issued_at_unix,
      expiresAtUnix: input.expires_at_unix,
      payloadDigest: input.payload_digest,
    };

    const bytes = marshalRelayEnvelope(envelope);
    expect(toHex(bytes)).toBe(vectors.relay_envelope_v1.canonical_hex);

    const roundTrip = parseRelayEnvelope(bytes);
    expect(roundTrip.requestId).toBe(input.request_id);
    expect(roundTrip.payloadDigest).toBe(input.payload_digest);
    expect(String(roundTrip.sequence)).toBe(String(input.sequence));
  });
});
